Q1: Are the majority of the CpG dinucleotides methylated or unmethylated
The majority of CpG dinucleotides are unmethylated.
 
ONT only: 6.604300474727729%
bisulfite only: 1.68827943122482%
both: 91.70742009404745%

Q2: How does using nanopore for methylation calling differ from bisulfite sequencing in terms of coverage? Which method appears better and why?

Bisulfite seqeuncing seems to have better coverage as around twice as much as ONT however the coverage also varies more and has a bigger distribution than ONT.

Q3: What can you infer about the two different approaches and their ability to detect methylation changes? 
The two methods seem similar in their ability to detect changes.

Q4: What is the effect of tumorigenesis on global methylation patterns?
Tumorigenesis seems to generally to not have a significant effect on methylation patterns but it may slightly tends towards less methylation globally.

Q5: What changes can you observe between the normal and tumor methylation landscape? What do you think the possible effects are of the changes you observed?
Tumor appears less methlylated. This makes sense since genes are improperly expressed in a disregulated way in cancer so it makes sense for there to be less methylation on DNMT3A.

Q6: What does it mean for a gene to be “imprinted”? 
Gene imprinting means that one copy of a gene (from one parent) is silenced and one (from the other parent) is expressed.

Q7: What is happening when you select the option to phase the reads? What is required in order to phase the reads?
Phasing reads allows us to see which allele the read is from. In order to phase reads there must be variants in the window of selection.

Q8: Can any set of reads be phased? Explain your answer.
Yes, you can see allele specific methylation sites. Tumorigenesis supresses more of these sites meaning there is less variation. No not every set of reads can be phased, there must be enough contextual information of other variants around.




